src/Buffer.mo

```diff
@@ -1,166 +1,1751 @@
-/// A buffer object can be extended by a single element or the contents of another buffer object.
+/// Throughout this documentation, two terms come up that can be confused: `size`
+/// and `capacity`. `size` is the length of the list that the buffer represents.
+/// `capacity` is the length of the underyling array that backs this list.
+/// `capacity` >= `size` is an invariant for this class.
+/// 
+/// Like arrays, elements in the buffer are ordered by indices from 0 to size-1.
 ///
-/// When required, the current state of a buffer object can be converted to a fixed-size array of its elements.
+/// WARNING: Certain operations are amortized O(1) time, such as `add`, but run
+/// in worst case O(n) time. These worst case runtimes may exceed the cycles limit
+/// per message if the size of the buffer is large enough. Grow these structures
+/// with discretion. All amortized operations below also list the worst case runtime.
+///
+/// Constructor:
+/// The argument `initCapacity` determines the initial capacity of the array.
+/// The underlying array grows by a factor of 1.5 when its current capacity is
+/// exceeded. Further, when the size of the buffer shrinks to be less than 1/4th
+/// of the capacity, the underyling array is shrunk by a factor of 2.
+///
+/// Runtime: O(initCapacity)
 ///
-/// Buffers complement Motoko's non-extensible array types
-/// (arrays do not support efficient extension, because the size of an array is
-/// determined at construction and cannot be changed).
+/// Space: O(initCapacity)
 
 import Prim "mo:⛔";
+import Result "Result";
+import Order "Order";
+import Array "Array";
 
 module {
+  type Order = Order.Order;
 
-  /// Create a stateful buffer class encapsulating a mutable array.
-  ///
-  /// The argument `initCapacity` determines its initial capacity.
-  /// The underlying mutable array grows by doubling when its current
-  /// capacity is exceeded.
-  public class Buffer<X>(initCapacity : Nat) {
-    var count : Nat = 0;
-    var elems : [var X] = [var]; // initially empty; allocated upon first `add`
-
-    /// Adds a single element to the buffer.
-    public func add(elem : X) {
-      if (count == elems.size()) {
-        let size =
-          if (count == 0) {
-            if (initCapacity > 0) { initCapacity } else { 1 }
-          } else {
-            2 * elems.size()
-          };
-        let elems2 = Prim.Array_init<X>(size, elem);
-        var i = 0;
-        label l loop {
-          if (i >= count) break l;
-          elems2[i] := elems[i];
-          i += 1;
-        };
-        elems := elems2;
+  // The following constants are used to manage the capacity.
+  // The length of `elements` is increased by `INCREASE_FACTOR` when capacity is reached.
+  // The length of `elements` is decreased by `DECREASE_FACTOR` when capacity is strictly less than
+  // `DECREASE_THRESHOLD`.
+  private let INCREASE_FACTOR = 1.5; // Keep this factor low to minimize cycle limit problem
```

Comment:
The floats make me slightly nervous - could you use a ratio instead? I.e. Integer multiply by 3 and divide by 2 or some other numerator denominator pair?

Reply:
Yup, will change